### ISSUE TITLE
Calling methods in externs upon wrong object should warn

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#3048]
I was calling a method which was missing from my externs, and was debugging why the compiler wasn't picking this up:

Externs:
``` javascript
var ExternsProblem = {};

/** @constructor */
var SomethingElse = function() {};

SomethingElse.prototype.something = function() {};
```
Code:
``` javascript
ExternsProblem.something(); // should fail, doesn't
```

You can see that `something` is defined elsewhere in the externs, but not here!

[Debugger](https://closure-compiler-debugger.appspot.com/#input0%3D%250AExternsProblem.something()%253B%2520%252F%252F%2520should%2520fail%252C%2520doesn't%250A%250A%26input1%26conformanceConfig%26externs%3Dvar%2520ExternsProblem%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250Avar%2520SomethingElse%2520%253D%2520function()%2520%257B%257D%253B%250A%250ASomethingElse.prototype.something%2520%253D%2520function()%2520%257B%257D%253B%250A%26refasterjs-template%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26CHECK_CONSTANTS%3Dtrue%26CHECK_DEPRECATED%3Dtrue%26CHECK_ES5_STRICT%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CHECK_GLOBAL_THIS%3Dtrue%26CHECK_LINT%3Dtrue%26CHECK_MISSING_RETURN%3Dtrue%26CHECK_UNREACHABLE_CODE%3Dtrue%26CHECK_PROVIDES%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CHECK_REPORT_MISSING_OVERRIDE%3Dtrue%26CHECK_SUSPICIOUS_CODE%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26CHECK_VISIBILITY%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue)

---

Ticking `ENABLE_ALL_DIAGNOSTIC_GROUPS` highlighted the problem: 
```
input0:2: WARNING - could not determine the type of this expression
ExternsProblem.something(); // should fail, doesn't
^^^^^^^^^^^^^^
```
[Debugger](https://closure-compiler-debugger.appspot.com/#input0%3D%250AExternsProblem.something()%253B%2520%252F%252F%2520should%2520fail%252C%2520doesn't%250A%250A%26input1%26conformanceConfig%26externs%3Dvar%2520ExternsProblem%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250Avar%2520SomethingElse%2520%253D%2520function()%2520%257B%257D%253B%250A%250ASomethingElse.prototype.something%2520%253D%2520function()%2520%257B%257D%253B%250A%26refasterjs-template%26ENABLE_ALL_DIAGNOSTIC_GROUPS%3Dtrue%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26CHECK_CONSTANTS%3Dtrue%26CHECK_DEPRECATED%3Dtrue%26CHECK_ES5_STRICT%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CHECK_GLOBAL_THIS%3Dtrue%26CHECK_LINT%3Dtrue%26CHECK_MISSING_RETURN%3Dtrue%26CHECK_UNREACHABLE_CODE%3Dtrue%26CHECK_PROVIDES%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CHECK_REPORT_MISSING_OVERRIDE%3Dtrue%26CHECK_SUSPICIOUS_CODE%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26CHECK_VISIBILITY%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue)

This error is [`UNKNOWN_EXPR_TYPE`](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/TypeCheck.java#L236) which has been disabled for 5 years.

---

The solution turns out to be to add `@const` to the externs:

``` javascript
/** @const */
var ExternsProblem = {};
...
```

Ah. Now I see the warning I expected:
```
input0:2: WARNING - Property something never defined on ExternsProblem
ExternsProblem.something(); // should fail, doesn't
               ^^^^^^^^^
```

[Debugger](https://closure-compiler-debugger.appspot.com/#input0%3D%250AExternsProblem.something()%253B%2520%252F%252F%2520should%2520fail%252C%2520doesn't%250A%250A%26input1%26conformanceConfig%26externs%3D%252F**%2520%2540const%2520*%252F%250Avar%2520ExternsProblem%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540constructor%2520*%252F%250Avar%2520SomethingElse%2520%253D%2520function()%2520%257B%257D%253B%250A%250ASomethingElse.prototype.something%2520%253D%2520function()%2520%257B%257D%253B%250A%26refasterjs-template%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26CHECK_CONSTANTS%3Dtrue%26CHECK_DEPRECATED%3Dtrue%26CHECK_ES5_STRICT%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CHECK_GLOBAL_THIS%3Dtrue%26CHECK_LINT%3Dtrue%26CHECK_MISSING_RETURN%3Dtrue%26CHECK_UNREACHABLE_CODE%3Dtrue%26CHECK_PROVIDES%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CHECK_REPORT_MISSING_OVERRIDE%3Dtrue%26CHECK_SUSPICIOUS_CODE%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26CHECK_VISIBILITY%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue)

---

I gather that the compiler doesn't know the type of external global var `ExternsProblem` so never complains about any property access upon it. This makes sense.

However we need a warning to prevent developers from calling functions that simply don't exist (typo) or aren't documented. 

I'm guessing `JSC_UNKNOWN_EXPR_TYPE` was disabled 5 years ago because it's old. Perhaps there's another check that should be kicking in here and isn't. I'm going to explore forcibly enabling this warning in all my externs, but would appreciate knowing how severe you think this issue is, and whether it affects other cases *outside* of externs.

**Comment by ChadKillingsworth** [comment]
I think the real takeaway here is we should warn whenever we see a non-constant namespace declaration in externs. Or just assume they are always constant.

**Comment by concavelenz** [comment]
Enable "strictMissingProperties" it should warn.  Please report back if you have already done this or if this doesn't fix the problem.

**Comment by concavelenz** [comment]
or more general "strictCheckTypes"

**Comment by supersteves** [comment]
I can't recall exactly but I have played with those strict options before (I think one was suggested as alternative NTI when it was removed) and I think it generated millions of warnings for commonly used patterns like:
``` javascript
let a = {};
a.something = ...
a.somethingElse = ...
```
This is just from memory, I'll need to revisit this in more detail.

**Comment by concavelenz** [comment]
It will certainly generate warning for values typed as "Object" but it understands literals:

https://closure-compiler-debugger.appspot.com/#input0%3D%250Alet%2520a%2520%253D%2520%257B%257D%253B%250Aa.something%2520%253D%25201%253B%250Aa.somethingElse%2520%253D%25202%253B%250A%250Ause(a.something)%253B%250Ause(a.bsProp)%253B%250A%250A%250Alet%2520b%2520%253D%2520%257B%257D%253B%250Ab.bsProp%2520%253D%25202%253B%250A%250A%250A%252F**%2520%2540type%2520%257B!Object%257D%2520*%252F%2520var%2520x%2520%253D%2520fn()%253B%250Ause(x.something)%253B%250Ause(x.bsProp)%253B%250A%26input1%26conformanceConfig%26externs%3Dvar%2520fn%253B%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

**Comment by concavelenz** [comment]
Closing in favor of https://github.com/google/closure-compiler/issues/3085

Reopen if you have more specific information.
